modules/linear_boltzmann_solvers/executors/pi_keigen_smm.cc

```diff
@@ -0,0 +1,1212 @@
+      {
+        const auto num_face_nodes = cell_mapping.NumFaceNodes(f);
+        for (int fi = 0; fi < num_face_nodes; ++fi)
+        {
+          const auto i = cell_mapping.MapFaceNode(f, fi);
+          const auto imap = pwld.MapDOFLocal(cell, i);
+
+          if (bndry_factors_.count(imap) == 0)
+            bndry_factors_[imap].resize(num_groupsets, 0.0);
+
+          // Compute the diffusion boundary factor, or the quadrature integral of
+          // |\Omega \cdot \hat{n}| divided by the sum of the quadrature weights.
+          double val = 0.0;
+          for (int d = 0; d < num_gs_dirs; ++d)
+            val += quad->weights_[d] * std::fabs(quad->omegas_[d].Dot(face.normal_));
+          bndry_factors_[imap][gs] = val / wt_sum;
+        }
+        ++f;
+      } // for face
+    }   // for cell
+    ++gs;
+  } // for groupset
+}
+
+void
+PowerIterationKEigenSMM::TransferTransportToDiffusion(const std::vector<double>& input,
+                                                      std::vector<double>& output) const
+{
+  const auto& grid = lbs_solver_.Grid();
+  const auto& pwld = lbs_solver_.SpatialDiscretization();
+  const auto& diff_sd = diffusion_solver_->SpatialDiscretization();
+
+  const auto& phi_uk_man = lbs_solver_.UnknownManager();
+  const auto& diff_uk_man = diffusion_solver_->UnknownStructure();
+
+  const auto first_grp = front_gs_.groups_.front().id_;
+  const auto num_gs_groups = front_gs_.groups_.size();
+
+  const auto num_local_diff_dofs = ghosts_required_ ? diff_sd.GetNumLocalAndGhostDOFs(diff_uk_man)
+                                                    : diff_sd.GetNumLocalDOFs(diff_uk_man);
+
+  // Check the input vector
+  if (input.size() != pwld.GetNumLocalDOFs(phi_uk_man))
+    throw std::logic_error("Vector size mismatch. Expected an input vector of size " +
+                           std::to_string(pwld.GetNumLocalDOFs(phi_uk_man)) + ", but got a " +
+                           "vector of size " + std::to_string(input.size()) + ".");
+
+  // If diffusion is PWLC, then nodal averages should be transferred to the PWLC vector,
+  // otherwise, the result is a copy
+  std::vector<double> mapped_phi;
+  if (pwlc_ptr_)
+    mapped_phi = ComputeNodallyAveragedPWLDVector(input, pwld, diff_sd, phi_uk_man, ghost_info_);
+  else
+    mapped_phi = input;
+
+  // Go through the cells and transfer the data to the output vector
+  output.assign(num_local_diff_dofs, 0.0);
+  for (const auto& cell : grid.local_cells)
+  {
+    // This is the same for PWLC and PWLD, so only this is needed
+    const auto& cell_mapping = pwld.GetCellMapping(cell);
+
+    for (int i = 0; i < cell_mapping.NumNodes(); ++i)
+    {
+      // Get the diffusion and transport DoFs
+      const auto input_dof_map = pwld.MapDOFLocal(cell, i, phi_uk_man, 0, first_grp);
+      const auto output_dof_map = diff_sd.MapDOFLocal(cell, i, diff_uk_man, 0, 0);
+
+      // Now copy the data
+      double* output_ptr = &output[output_dof_map];
```

Comment:
I think:
```
std::copy_n(std::advance(mapped_phi.begin() + input_dof_map), 
            num_gs_groups,
            std::advance(output.begin() + output_dof_map));
```
will work here without having to devolve everything to pointers.

Reply:
How about something like: 
```
std::copy_n(&mapped_phi[input_dof_map], num_gs_groups, &output[output_dof_map]);
```
I think `std::advance` does the same thing as `+` in your suggestion, so `+` should be a `,` instead, or `std::advance` not used.

Reply:
You are right, the plus should be a comma in my code above. I think I originally was going for `mapped_phi.begin() + input_dof_map` which would also work. The `std::advance` just made it more general.  These are all vectors and probably won't change, so your `copy_n` solution is just fine.